packages/core/src/RenderingEngine/WSIViewport.ts

```diff
@@ -217,17 +230,63 @@ class WSIViewport extends Viewport {
   }
 
   public setProperties(props: WSIViewportProperties) {
-    // No-op - todo implement this
+    if (props.voiRange) {
+      this.setVOI(props.voiRange);
+    }
   }
 
   public getProperties = (): WSIViewportProperties => {
-    return {};
+    return {
+      voiRange: { ...this.voiRange },
+    };
   };
 
   public resetProperties() {
     this.setProperties({});
   }
 
+  public setVOI(voiRange: VOIRange): void {
+    // This sets the window level
+    this.voiRange = voiRange;
+    this.setColorTransform();
+  }
+
+  public setAverageWhite(averageWhite: [number, number, number]) {
+    this.averageWhite = averageWhite;
+    this.setColorTransform();
+  }
+
+  protected setColorTransform() {
```

Comment:
Can you move the setColorTransform call to Viewport, and just describe it as a filter on the canvas?  That way both Video and WSI Viewports use the same code.  The setColorTransform can take the voiRange/averageWhite as arguments.

Reply:
Stack/Volume should not be able to call them, i think we need another base class

Reply:
It doesn't matter if stack/volume calls them, it would apply as a secondary window level, and there are cases that might be worth allowing such as having a JPEG original image.  Applying window level using a transform for that case avoids certain types of issues around caching data.  However, as a protected method or even  a static method one could just declare it as being a helper method in the docs.  However, you might be right, adding a new base class for non-vtk based stacks might be worthwhile as there are other helper methods to consider.